### PR TITLE
Fixes most of the tests

### PR DESCRIPTION
**Pull request by dmitraver**
Hi! 

I've fixed most of the tests except LogCompactionTest and ClusterMembershipChangeTest. I also changed some logic in ReplicatedLog class that refers to some indexes computations. I hope I didn't violate any Raft constraints :) 

@colin-scott If possible it could be nice if you can also take a look.

I'm currently trying to fix and improve log compaction but there are some places in a raft paper which are not quite clear about the implementation but I hope to make a new PR for that soon.

### CODE REVIEW

File: src/main/scala/pl/project13/scala/akka/raft/protocol/StateMetadata.scala [hunk 1]
```diff
@@ -41,13 +41,17 @@ private[protocol] trait StateMetadata extends Serializable {
...
     // transition helpers
-    def forNewElection: ElectionMeta = ElectionMeta(clusterSelf, currentTerm.next, 0, config, votedFor)
+    def forNewElection: ElectionMeta = ElectionMeta(clusterSelf, currentTerm.next, 0, config, None)
 
-    def withTerm(term: Term) = copy(currentTerm = term)
+    def withTerm(term: Term) = {
+      if (term != currentTerm)
+        copy(currentTerm = term, votedFor = None)
+      else this
+    }
```
Comment (ktoso) [inline src/main/scala/pl/project13/scala/akka/raft/protocol/StateMetadata.scala]: +1, good

File: src/main/scala/pl/project13/scala/akka/raft/model/ReplicatedLog.scala [hunk 2]
```diff
@@ -19,15 +19,15 @@ case class ReplicatedLog[Command](
...
+    (otherPrevTerm == Term(0) && otherPrevIndex == 1) ||
+    (entries.isDefinedAt(otherPrevIndex - 1) && entries(otherPrevIndex - 1).term == otherPrevTerm && lastIndex == otherPrevIndex)
 
   // log state
   def lastTerm  = entries.lastOption map { _.term } getOrElse Term(0)
   def lastIndex = entries.lastOption map { _.index } getOrElse 1
 
   def prevIndex = (lastIndex: @switch) match {
-    case 0 => 0 // special handling of initial case, we don't go into negative indexes
+    case 1 => 1 // special handling of initial case, we don't go into negative indexes
```
Comment (ktoso) [inline src/main/scala/pl/project13/scala/akka/raft/model/ReplicatedLog.scala]: 👍

**Comment by ktoso** [comment]
Thanks a lot for keeping at it :-)

**Comment by dmitraver** [comment]
Thanks! I will try to fix the log compaction cause I want tests to be working again :)

**Pull request merged**